From: =?UTF-8?q?Mika=20Pfl=C3=BCger?= <debian@mikapflueger.de>
Date: Mon, 27 Feb 2012 01:36:07 +0100
Subject: Add debian apache paths and apache_script_exec_domain()

The patch firstly adds support for some Debian paths to the Apache
policy and has some minor changes to the access granted to user content.

The biggest change is to have a new interface apache_script_exec_domain() for
creating types for Apache content which doesn't need to be optional.  So the
types created by it can be used in a .fc file which will be included even if
apache.pp isn't loaded.  This means that a module which has a script that
Apache might run won't depend on apache.pp.
---
 policy/modules/apps/awstats.te     |    1 +
 policy/modules/services/apache.fc  |    7 ++++++-
 policy/modules/services/apache.if  |   35 +++++++++++++++++++++++++++++++++--
 policy/modules/services/apache.te  |   34 +++++++++++++++++++++++++++++++++-
 policy/modules/services/apcupsd.te |    1 +
 policy/modules/services/cvs.te     |    1 +
 policy/modules/services/git.te     |    1 +
 policy/modules/services/munin.te   |    1 +
 policy/modules/services/prelude.te |    1 +
 policy/modules/services/squid.te   |    1 +
 policy/modules/services/w3c.te     |    1 +
 11 files changed, 80 insertions(+), 4 deletions(-)

diff --git a/policy/modules/apps/awstats.te b/policy/modules/apps/awstats.te
index 6bd3ad3..eb2004d 100644
--- a/policy/modules/apps/awstats.te
+++ b/policy/modules/apps/awstats.te
@@ -17,6 +17,7 @@ files_tmp_file(awstats_tmp_t)
 type awstats_var_lib_t;
 files_type(awstats_var_lib_t)
 
+apache_script_exec_domain(awstats)
 apache_content_template(awstats)
 
 ########################################
diff --git a/policy/modules/services/apache.fc b/policy/modules/services/apache.fc
index 9e39aa5..a9eb531 100644
--- a/policy/modules/services/apache.fc
+++ b/policy/modules/services/apache.fc
@@ -32,8 +32,12 @@ HOME_DIR/((www)|(web)|(public_html))(/.+)? gen_context(system_u:object_r:httpd_u
 /usr/lib(64)?/httpd(/.*)?		gen_context(system_u:object_r:httpd_modules_t,s0)
 /usr/lib(64)?/lighttpd(/.*)?		gen_context(system_u:object_r:httpd_modules_t,s0)
 
+ifdef(`distro_debian', `
+/usr/lib/apache2/mpm-.*/.*$	--	gen_context(system_u:object_r:httpd_exec_t,s0)
+', `
 /usr/sbin/apache(2)?		--	gen_context(system_u:object_r:httpd_exec_t,s0)
 /usr/sbin/apache-ssl(2)?	--	gen_context(system_u:object_r:httpd_exec_t,s0)
+')
 /usr/sbin/httpd(\.worker)?	--	gen_context(system_u:object_r:httpd_exec_t,s0)
 /usr/sbin/lighttpd		--	gen_context(system_u:object_r:httpd_exec_t,s0)
 /usr/sbin/rotatelogs		--	gen_context(system_u:object_r:httpd_rotatelogs_exec_t,s0)
@@ -79,7 +83,8 @@ ifdef(`distro_suse', `
 /var/lib/httpd(/.*)?			gen_context(system_u:object_r:httpd_var_lib_t,s0)
 /var/lib/php/session(/.*)?		gen_context(system_u:object_r:httpd_var_run_t,s0)
 /var/lib/squirrelmail/prefs(/.*)?	gen_context(system_u:object_r:httpd_squirrelmail_t,s0)
-
+/var/lib/squirrelmail/data(/.*)?        gen_context(system_u:object_r:httpd_squirrelmail_t,s0)
+ 
 /var/log/apache(2)?(/.*)?		gen_context(system_u:object_r:httpd_log_t,s0)
 /var/log/apache-ssl(2)?(/.*)?		gen_context(system_u:object_r:httpd_log_t,s0)
 /var/log/cacti(/.*)?			gen_context(system_u:object_r:httpd_log_t,s0)
diff --git a/policy/modules/services/apache.if b/policy/modules/services/apache.if
index 6480167..9667d8c 100644
--- a/policy/modules/services/apache.if
+++ b/policy/modules/services/apache.if
@@ -11,12 +11,29 @@
 ##	</summary>
 ## </param>
 #
+template(`apache_script_exec_domain',`
+	type httpd_$1_script_exec_t; # customizable;
+	fs_associate(httpd_$1_script_exec_t)
+')
+
+########################################
+## <summary>
+##	Create a set of derived types for apache
+##	web content.
+## </summary>
+## <param name="prefix">
+##	<summary>
+##	The prefix to be used for deriving type names.
+##	</summary>
+## </param>
+#
 template(`apache_content_template',`
 	gen_require(`
 		attribute httpdcontent;
 		attribute httpd_exec_scripts;
 		attribute httpd_script_exec_type;
 		type httpd_t, httpd_suexec_t, httpd_log_t;
+		type httpd_$1_script_exec_t;
 	')
 	# allow write access to public file transfer
 	# services files.
@@ -37,7 +54,9 @@ template(`apache_content_template',`
 	role system_r types httpd_$1_script_t;
 
 	# This type is used for executable scripts files
-	type httpd_$1_script_exec_t, httpd_script_exec_type; # customizable;
+	# must be defined by the caller
+	# type httpd_$1_script_exec_t, httpd_script_exec_type; # customizable;
+	typeattribute httpd_$1_script_exec_t httpd_script_exec_type;
 	corecmd_shell_entry_type(httpd_$1_script_t)
 	domain_entry_file(httpd_$1_script_t, httpd_$1_script_exec_t)
 
@@ -50,6 +69,7 @@ template(`apache_content_template',`
 	files_type(httpd_$1_ra_content_t)
 
 	read_files_pattern(httpd_t, httpd_$1_content_t, httpd_$1_htaccess_t)
+	read_lnk_files_pattern(httpd_t, httpd_$1_content_t, httpd_$1_htaccess_t)
 
 	domtrans_pattern(httpd_suexec_t, httpd_$1_script_exec_t, httpd_$1_script_t)
 
@@ -108,6 +128,10 @@ template(`apache_content_template',`
 
 	seutil_dontaudit_search_config(httpd_$1_script_t)
 
+	allow httpd_t httpd_$1_content_t:dir list_dir_perms;
+	read_files_pattern(httpd_t,httpd_$1_content_t,httpd_$1_content_t)
+	read_lnk_files_pattern(httpd_t,httpd_$1_content_t,httpd_$1_content_t)
+
 	tunable_policy(`httpd_enable_cgi && httpd_unified',`
 		allow httpd_$1_script_t httpdcontent:file entrypoint;
 
@@ -121,7 +145,7 @@ template(`apache_content_template',`
 		miscfiles_manage_public_files(httpd_$1_script_t)
 	')
 
-	# Allow the web server to run scripts and serve pages
+	# Allow the web server to run scripts
 	tunable_policy(`httpd_builtin_scripting',`
 		manage_dirs_pattern(httpd_t, httpd_$1_rw_content_t, httpd_$1_rw_content_t)
 		manage_files_pattern(httpd_t, httpd_$1_rw_content_t, httpd_$1_rw_content_t)
@@ -227,6 +251,13 @@ interface(`apache_role',`
 	relabel_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
 	relabel_lnk_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
 
+	manage_dirs_pattern($2, httpd_user_content_t, httpd_user_content_t)
+	manage_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
+	manage_lnk_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
+	relabel_dirs_pattern($2, httpd_user_content_t, httpd_user_content_t)
+	relabel_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
+	relabel_lnk_files_pattern($2, httpd_user_content_t, httpd_user_content_t)
+
 	manage_dirs_pattern($2, httpd_user_ra_content_t, httpd_user_ra_content_t)
 	manage_files_pattern($2, httpd_user_ra_content_t, httpd_user_ra_content_t)
 	manage_lnk_files_pattern($2, httpd_user_ra_content_t, httpd_user_ra_content_t)
diff --git a/policy/modules/services/apache.te b/policy/modules/services/apache.te
index 5b02edb..18843d1 100644
--- a/policy/modules/services/apache.te
+++ b/policy/modules/services/apache.te
@@ -215,6 +215,7 @@ type httpd_suexec_tmp_t;
 files_tmp_file(httpd_suexec_tmp_t)
 
 # setup the system domain for system CGI scripts
+apache_script_exec_domain(sys)
 apache_content_template(sys)
 typealias httpd_sys_content_t alias ntop_http_content_t;
 
@@ -224,6 +225,7 @@ files_tmp_file(httpd_tmp_t)
 type httpd_tmpfs_t;
 files_tmpfs_file(httpd_tmpfs_t)
 
+apache_script_exec_domain(user)
 apache_content_template(user)
 ubac_constrained(httpd_user_script_t)
 userdom_user_home_content(httpd_user_content_t)
@@ -469,6 +471,14 @@ tunable_policy(`httpd_enable_ftp_server',`
 tunable_policy(`httpd_enable_homedirs',`
 	userdom_read_user_home_content_files(httpd_t)
 ')
+optional_policy(`
+	gen_require(`
+		bool daemon_access_unconfined_home;
+	')
+	tunable_policy(`httpd_enable_homedirs && daemon_access_unconfined_home', `
+			unconfined_read_home_content_files(httpd_t)
+	')
+')
 
 tunable_policy(`httpd_enable_homedirs && use_nfs_home_dirs',`
 	fs_read_nfs_files(httpd_t)
@@ -517,7 +527,10 @@ optional_policy(`
 ')
 
 optional_policy(`
+# for cron jobs to restart Apache
 	cron_system_entry(httpd_t, httpd_exec_t)
+# For cron jobs to run from accounts with home directories in the web store
+	crond_search_dir(httpd_sys_content_t)
 ')
 
 optional_policy(`
@@ -542,8 +555,11 @@ optional_policy(`
 	')
 ')
 
-optional_policy(`
 	kerberos_keytab_template(httpd, httpd_t)
+
+optional_policy(`
+        # read munin files
+        munin_search_lib(httpd_t)
 ')
 
 optional_policy(`
@@ -739,6 +755,14 @@ tunable_policy(`httpd_can_network_connect',`
 	corenet_tcp_connect_all_ports(httpd_suexec_t)
 	corenet_sendrecv_all_client_packets(httpd_suexec_t)
 ')
+optional_policy(`
+	gen_require(`
+		bool daemon_access_unconfined_home;
+	')
+	tunable_policy(`httpd_enable_homedirs && daemon_access_unconfined_home', `
+			unconfined_read_home_content_files(httpd_suexec_t)
+	')
+')
 
 tunable_policy(`httpd_enable_cgi && httpd_unified',`
 	allow httpd_sys_script_t httpdcontent:file entrypoint;
@@ -824,6 +848,14 @@ tunable_policy(`httpd_enable_cgi && httpd_can_network_connect',`
 tunable_policy(`httpd_enable_homedirs',`
 	userdom_read_user_home_content_files(httpd_sys_script_t)
 ')
+optional_policy(`
+	gen_require(`
+		bool daemon_access_unconfined_home;
+	')
+	tunable_policy(`httpd_enable_homedirs && daemon_access_unconfined_home', `
+			unconfined_read_home_content_files(httpd_sys_script_t)
+	')
+')
 
 tunable_policy(`httpd_enable_homedirs && use_nfs_home_dirs',`
 	fs_read_nfs_files(httpd_sys_script_t)
diff --git a/policy/modules/services/apcupsd.te b/policy/modules/services/apcupsd.te
index d052bf0..00d1845 100644
--- a/policy/modules/services/apcupsd.te
+++ b/policy/modules/services/apcupsd.te
@@ -107,6 +107,7 @@ optional_policy(`
 # apcupsd_cgi Declarations
 #
 
+apache_script_exec_domain(apcupsd_cgi)
 optional_policy(`
 	apache_content_template(apcupsd_cgi)
 
diff --git a/policy/modules/services/cvs.te b/policy/modules/services/cvs.te
index 88e7e97..7ecf618 100644
--- a/policy/modules/services/cvs.te
+++ b/policy/modules/services/cvs.te
@@ -106,6 +106,7 @@ optional_policy(`
 # CVSWeb policy
 #
 
+apache_script_exec_domain(cvs)
 optional_policy(`
 	apache_content_template(cvs)
 
diff --git a/policy/modules/services/git.te b/policy/modules/services/git.te
index 7382f85..5f6ec99 100644
--- a/policy/modules/services/git.te
+++ b/policy/modules/services/git.te
@@ -5,4 +5,5 @@ policy_module(git, 1.0)
 # Declarations
 #
 
+apache_script_exec_domain(git)
 apache_content_template(git)
diff --git a/policy/modules/services/munin.te b/policy/modules/services/munin.te
index f17583b..4d6d86d 100644
--- a/policy/modules/services/munin.te
+++ b/policy/modules/services/munin.te
@@ -122,6 +122,7 @@ sysnet_exec_ifconfig(munin_t)
 userdom_dontaudit_use_unpriv_user_fds(munin_t)
 userdom_dontaudit_search_user_home_dirs(munin_t)
 
+apache_script_exec_domain(munin)
 optional_policy(`
 	apache_content_template(munin)
 
diff --git a/policy/modules/services/prelude.te b/policy/modules/services/prelude.te
index b1bc02c..d565828 100644
--- a/policy/modules/services/prelude.te
+++ b/policy/modules/services/prelude.te
@@ -278,6 +278,7 @@ optional_policy(`
 # prewikka_cgi Declarations
 #
 
+apache_script_exec_domain(prewikka)
 optional_policy(`
 	apache_content_template(prewikka)
 
diff --git a/policy/modules/services/squid.te b/policy/modules/services/squid.te
index 4b2230e..26dc155 100644
--- a/policy/modules/services/squid.te
+++ b/policy/modules/services/squid.te
@@ -177,6 +177,7 @@ tunable_policy(`squid_use_tproxy',`
 	corenet_tcp_bind_netport_port(squid_t)
 ')
 
+apache_script_exec_domain(squid)
 optional_policy(`
 	apache_content_template(squid)
 
diff --git a/policy/modules/services/w3c.te b/policy/modules/services/w3c.te
index 1174ad8..05a34ee 100644
--- a/policy/modules/services/w3c.te
+++ b/policy/modules/services/w3c.te
@@ -5,6 +5,7 @@ policy_module(w3c, 1.0.0)
 # Declarations
 #
 
+apache_script_exec_domain(w3c_validator)
 apache_content_template(w3c_validator)
 
 ########################################
